internal/types/path.go

```diff
@@ -156,40 +156,119 @@ func RemoveByPath[T CompositeTypeInterface](comp T, path Path) {
 	removeByPath(comp, path)
 }
 
-// getByPath returns a value by path - a sequence of indexes and keys.
-func getByPath[T CompositeTypeInterface](comp T, path Path) (any, error) {
+// getExactByPath returns all matching values by path - a sequence of indexes and keys.
+func getExactByPath[T CompositeTypeInterface](comp T, path Path) (any, error) {
 	var next any = comp
 	for _, p := range path.Slice() {
 		switch s := next.(type) {
 		case *Document:
 			var err error
 			next, err = s.Get(p)
 			if err != nil {
-				return nil, newDocumentPathError(ErrDocumentPathKeyNotFound, fmt.Errorf("types.getByPath: %w", err))
+				return nil, newDocumentPathError(ErrDocumentPathKeyNotFound, fmt.Errorf("types.getExactByPath: %w", err))
 			}
 
 		case *Array:
 			index, err := strconv.Atoi(p)
 			if err != nil {
-				return nil, newDocumentPathError(ErrDocumentPathArrayInvalidIndex, fmt.Errorf("types.getByPath: %w", err))
+				return nil, newDocumentPathError(ErrDocumentPathArrayInvalidIndex, fmt.Errorf("types.getExactByPath: %w", err))
 			}
 
 			next, err = s.Get(index)
 			if err != nil {
-				return nil, newDocumentPathError(ErrDocumentPathIndexOutOfBound, fmt.Errorf("types.getByPath: %w", err))
+				return nil, newDocumentPathError(ErrDocumentPathIndexOutOfBound, fmt.Errorf("types.getExactByPath: %w", err))
 			}
 
 		default:
 			return nil, newDocumentPathError(
 				ErrDocumentPathCannotAccess,
-				fmt.Errorf("types.getByPath: can't access %T by path %q", next, p),
+				fmt.Errorf("types.getExactByPath: can't access %T by path %q", next, p),
 			)
 		}
 	}
 
 	return next, nil
 }
 
+// getAllByPath returns all matching values by path - a sequence of indexes and keys.
+func getAllByPath[T CompositeTypeInterface](comp T, path Path) ([]any, error) {
+	next := []any{comp}
+
+	for _, p := range path.Slice() {
+		var newNext []any
```

Comment:
Just a suggestion, doesn't need to be applied but I feel this can be read easier if we have `current` and `next` :innocent: